src/app/digital/utils/ExpressionParser/GenerateInputTree.ts

```diff
@@ -0,0 +1,144 @@
+import {Token, FormatLabels, TokenType, NewTreeRetValue, InputTree} from "./Constants/DataStructures";
+import {DefaultPrecedences, FormatMap} from "./Constants/Objects";
+import {IsOperator} from "./Utils";
+
+
+/**
+ * The core of the function to generate the input tree. Various errors are returned for invalid inputs.
+ *  It is recommended to not call this function directly and instead call GenerateInputTree
+ * 
+ * @param tokens the array of tokens representing the expression to parse
+ * @param currentOp the current operation to evaluate, should default to operation with lowest precedence to start
+ * @param index the index of the parsing process in the tokens Array
+ * @returns the current input tree and the current parsing index
+ * @throws {Error} parenthesis do not include anything (such as "()")
+ * @throws {Error} an opening parenthesis is missing a corresponding closing parenthesis (such as "(a")
+ * @throws {Error} a closing parenthesis is missing a corresponding opening parenthesis (such as ")a")
+ * @throws {Error} |, &, or ^ are missing an operand on their left (such as "a|")
+ * @throws {Error} |, &, ^, or ! are missing an operand on their right (such as "!a")
+ * @see GenerateInputTree
+ */
+ function generateInputTreeCore(tokens: Array<Token>, ops: Map<FormatLabels, string>, currentOp: TokenType = "|", index: number = 0): NewTreeRetValue {
+    const nextOp = DefaultPrecedences.get(currentOp);
+    if (tokens[index].type === ")") {
+        if (index > 0) {
+            if (tokens[index-1].type === "(")
+                throw new Error("Empty Parenthesis");
+            else if (IsOperator(tokens[index-1]))
```

Comment:
nit: this can just be another `if` since the previous `if` throws and therefore stops execution